luci-app-squid: convert to JavaScript

dannil opened the pull request:
The conversion adds the missing parameters in the UI which is configurable in the principal package.

Assumption has been made that the config file is stored in /etc/squid since the principal package limits the sysconfdir to this directory. If that assumption is changed in the future we need to adjust the ACL.

![Screenshot_1](https://github.com/user-attachments/assets/e3ac3718-f022-4751-aaa4-6eb4b4095eac)
![Screenshot_2](https://github.com/user-attachments/assets/86ce98df-b67d-49af-a21c-801788535137)
![Screenshot_3](https://github.com/user-attachments/assets/2bf050ae-f56f-404e-8023-b3394141a48c)

Values are updated appropriately which the squid logs also indicates.

```
root@labmouse:~# uci show squid
squid.squid=squid
squid.squid.config_file='/etc/squid/squid.conf'
squid.squid.http_port='12345'
squid.squid.coredump_dir='/tmp/customsquid'
squid.squid.visible_hostname='foobar'
squid.squid.mime_table='/etc/squid/mime.conf'
squid.squid.pinger_enable='off'
```

```
Thu Oct 17 18:03:13 2024 daemon.notice squid[4163]: Processing Configuration File: /tmp/squid/squid.conf (depth 0)
Thu Oct 17 18:03:13 2024 daemon.notice squid[4163]: ERROR: 'pinger_enable' requires --enable-icmp
Thu Oct 17 18:03:13 2024 daemon.warn squid[4163]: FATAL: Squid is already running: Found fresh instance PID file (/var/run/squid.pid) with PID 4095     exception location: Instance.cc(122) ThrowIfAlreadyRunningWith
Thu Oct 17 18:03:13 2024 daemon.notice squid[4095]: Preparing for shutdown after 0 requests
Thu Oct 17 18:03:13 2024 daemon.notice squid[4095]: Waiting 30 seconds for active connections to finish
Thu Oct 17 18:03:13 2024 daemon.notice squid[4095]: Closing HTTP(S) port [::]:1337     listening port: 1337
Thu Oct 17 18:03:18 2024 daemon.info procd: Instance squid::instance1 pid 4095 not stopped on SIGTERM, sending SIGKILL instead
Thu Oct 17 18:03:18 2024 daemon.notice squid[4220]: Processing Configuration File: /tmp/squid/squid.conf (depth 0)
Thu Oct 17 18:03:18 2024 daemon.notice squid[4220]: ERROR: 'pinger_enable' requires --enable-icmp
Thu Oct 17 18:03:18 2024 daemon.notice squid[4220]: Created PID file (/var/run/squid.pid)
Thu Oct 17 18:03:18 2024 daemon.notice squid[4220]: Set Current Directory to /tmp/customsquid
Thu Oct 17 18:03:18 2024 daemon.warn squid[4220]: Starting Squid Cache version 6.11 for x86_64-openwrt-linux-gnu...
Thu Oct 17 18:03:18 2024 daemon.warn squid[4220]: Service Name: squid
Thu Oct 17 18:03:18 2024 daemon.notice squid[4220]: Process ID 4220
Thu Oct 17 18:03:18 2024 daemon.notice squid[4220]: Process Roles: master worker
Thu Oct 17 18:03:18 2024 daemon.notice squid[4220]: With 1024 file descriptors available
Thu Oct 17 18:03:18 2024 daemon.notice squid[4220]: Initializing IP Cache...
Thu Oct 17 18:03:18 2024 daemon.notice squid[4220]: DNS IPv6 socket created at [::], FD 9
Thu Oct 17 18:03:18 2024 daemon.notice squid[4220]: DNS IPv4 socket created at 0.0.0.0, FD 10
Thu Oct 17 18:03:18 2024 daemon.notice squid[4220]: Adding domain lan from /etc/resolv.conf
Thu Oct 17 18:03:18 2024 daemon.notice squid[4220]: Adding nameserver 127.0.0.1 from /etc/resolv.conf
Thu Oct 17 18:03:18 2024 daemon.notice squid[4220]: Adding nameserver ::1 from /etc/resolv.conf
Thu Oct 17 18:03:18 2024 daemon.notice squid[4220]: Local cache digest enabled; rebuild/rewrite every 3600/3600 sec
Thu Oct 17 18:03:18 2024 daemon.notice squid[4220]: Store logging disabled
Thu Oct 17 18:03:18 2024 daemon.notice squid[4220]: Swap maxSize 0 + 262144 KB, estimated 20164 objects
Thu Oct 17 18:03:18 2024 daemon.notice squid[4220]: Target number of buckets: 1008
Thu Oct 17 18:03:18 2024 daemon.notice squid[4220]: Using 8192 Store buckets
Thu Oct 17 18:03:18 2024 daemon.notice squid[4220]: Max Mem  size: 262144 KB
Thu Oct 17 18:03:18 2024 daemon.notice squid[4220]: Max Swap size: 0 KB
Thu Oct 17 18:03:18 2024 daemon.notice squid[4220]: Using Least Load store dir selection
Thu Oct 17 18:03:18 2024 daemon.notice squid[4220]: Set Current Directory to /tmp/customsquid
Thu Oct 17 18:03:18 2024 daemon.notice squid[4220]: Finished loading MIME types and icons.
Thu Oct 17 18:03:18 2024 daemon.notice squid[4220]: HTCP Disabled.
Thu Oct 17 18:03:18 2024 daemon.notice squid[4220]: Squid plugin modules loaded: 0
Thu Oct 17 18:03:18 2024 daemon.notice squid[4220]: Accepting HTTP Socket connections at conn3 local=[::]:12345 remote=[::] FD 11 flags=9     listening port: 12345
Thu Oct 17 18:03:19 2024 daemon.notice squid[4220]: storeLateRelease: released 0 objects
```

Updating the files via advanced settings also works.

![Screenshot_4](https://github.com/user-attachments/assets/76fa1c7e-e67c-4e89-9bc3-e88b4c99e0b9)
![Screenshot_5](https://github.com/user-attachments/assets/b1dcc09a-5bde-4191-974e-d43a326bce65)

```
root@labmouse:~# cat /etc/squid/squid.conf
# this is the config file

root@labmouse:~# cat /etc/squid/mime.conf
# this is the mime table
```

As discussed in https://github.com/openwrt/luci/pull/7320#issuecomment-2408060031, if squid is compiled with --enable-icmp, the option will be enabled, otherwise it'll be disabled with a tool tip explaining why.

![Screenshot_6](https://github.com/user-attachments/assets/ab3f8256-9865-4b8d-8204-7e2a2d58ce92)
![Screenshot_7](https://github.com/user-attachments/assets/8190a75f-a368-48f8-ba23-9623be2b8fe4)

- [X] This PR is not from my *main* or *master* branch :poop:, but a *separate* branch :white_check_mark:
- [X] Each commit has a valid :black_nib: `Signed-off-by: <my@email.address>` row (via `git commit --signoff`)
- [X] Each commit and PR title has a valid :memo: `<package name>: title` first line subject for packages
- [X] Incremented :up: any `PKG_VERSION` in the Makefile N/A
- [X] Tested on: (x86/64, SNAPSHOT r27707-084665698b, Firefox) :white_check_mark:
- [X] \( Preferred ) Mention: @ the original code author for feedback @ratkaj
- [X] \( Preferred ) Screenshot or mp4 of changes:
- [ ] \( Optional ) Closes: e.g. openwrt/luci#issue-number
- [ ] \( Optional ) Depends on: e.g. openwrt/packages#pr-number in sister repo
- [X] Description: (describe the changes proposed in this PR)

systemcrash commented:
This looks fine as-is. If you're feeling up to it, could you also add the extra variables found in the squid package please ( as at 23.05 ):

```sh
validate_squid_section() {
	uci_load_validate squid squid "$1" "$2" \
		'config_file:string' \
		'http_port:port:3128' \
		'http_port_options:string' \
		'ssldb:string' \
		'ssldb_options:string' \
		'coredump_dir:string' \
		'visible_hostname:string:OpenWrt' \
		'pinger_enable:string:off' \
		'mime_table:string:/etc/squid/mime.conf'
}

```

dannil commented:
> This looks fine as-is. If you're feeling up to it, could you also add the extra variables found in the squid package please ( as at 23.05 ):
> 
> ```shell
> validate_squid_section() {
> 	uci_load_validate squid squid "$1" "$2" \
> 		'config_file:string' \
> 		'http_port:port:3128' \
> 		'http_port_options:string' \
> 		'ssldb:string' \
> 		'ssldb_options:string' \
> 		'coredump_dir:string' \
> 		'visible_hostname:string:OpenWrt' \
> 		'pinger_enable:string:off' \
> 		'mime_table:string:/etc/squid/mime.conf'
> }
> ```

Sure thing, should be very straight-forward.

dannil commented:
I've basically got everything working with the new options, but during this I discovered that [the pinger_enable option requires Squid to be started with the --enable-icmp parameter](http://www.squid-cache.org/Doc/config/pinger_enable/), which the principal package doesn't do. This can also be seen in the logs:

```
Fri Oct 11 20:01:30 2024 daemon.notice squid[4620]: ERROR: 'pinger_enable' requires --enable-icmp
```

Should we still expose this option even though it won't do anything?

hnyman commented:
LuCI should match the principal package.
It makes no sense to offer options that the package does not actually support.

systemcrash commented:
Is it not what the start script makes possible?

dannil commented:
> Is it not what the start script makes possible?

A prerequisite of being able to use `pinger_enable` (even if you set it `on` or `off` doesn't matter, the warning will show up irregardless otherwise) is that Squid has been compiled with the `--enable-icmp` option, which isn't enabled in the config at https://github.com/openwrt/packages/blob/552f561c51abac603cc40229f3cfa89e32ef3127/net/squid/Config.in#L13-L15 which gets passed through at https://github.com/openwrt/packages/blob/552f561c51abac603cc40229f3cfa89e32ef3127/net/squid/Makefile#L100. So for everyone which doesn't set this in menuconfig (which is everyone not compiling from source) this option won't do anything.

Also, another interesting thing is that the fallback compile option then becomes `--disable-icmp`, which as of Squid 6.11 isn't even a valid parameter, so I guess it gets silently ignored.

systemcrash commented:
Ah. Okay. Previously you wrote started. I started to suspect it was a CTO.  Technically it's possible to parse the help output of squid to scrape whether it's compiled so, but that's another problem for another day. Skip ping 👌

dannil commented:
> Ah. Okay. Previously you wrote started. I started to suspect it was a CTO. Technically it's possible to parse the help output of squid to scrape whether it's compiled so, but that's another problem for another day. Skip ping 👌

Sounds good. Yeah I see the misunderstanding, I had actually missed it was a compile option, not just a parameter, my bad.

hnyman commented:
If it is a compile-time option, one alternative might be to add a note to the option explanation in LuCI, that it a compile option is required for the functionality to work.

systemcrash commented:
You can take a look at how we did it in (luci-app-)lldpd.

dannil commented:
FYI: I'm gonna remove the custom `handleSaveApply` implementation as I noticed that squid [sets up a procd reload trigger](https://github.com/openwrt/packages/blob/552f561c51abac603cc40229f3cfa89e32ef3127/net/squid/files/squid.init#L72) for it's UCI section, so migrating the `on_after_commit` function in the Lua implementation isn't even necessary. The more you know!

systemcrash commented:
How's it coming with `handleSaveApply`? Do you prefer to add GUI restart helpers and let uci handle the reload?

dannil commented:
> How's it coming with `handleSaveApply`? Do you prefer to add GUI restart helpers and let uci handle the reload?

I basically have the entire PR sorted out on my end, just haven't had the time to finalize it. Since the principal package sets up a procd trigger on the uci squid section, there's no easy way for us anyway to handle the reload manually on the LuCI app with restart helpers, so I'm just gonna leave it the way it was before (as in the default `Save and Apply` reloading squid), which works as it should.

dannil commented:
PR should be good to go now. Changes since putting in draft:

* Added missing options according to https://github.com/openwrt/luci/pull/7320#issuecomment-2405904435
* `Enable ICMP pinger` is conditionally enabled depending on compile-time options: https://github.com/openwrt/luci/pull/7320#issuecomment-2408060031
* Moved config file and mime table to separate tabs under `Advanced Settings`, it started to feel very cluttered to try and mash that into the same page with subheadings.

systemcrash commented on applications/luci-app-squid/htdocs/luci-static/resources/view/squid.js:
```diff
@@ -0,0 +1,134 @@
+		o.datatype = 'string';
+		o.placeholder = '/tmp/squid';
+
+		var enable_icmp_option = '--enable-icmp';
```
```js
		var enable_icmp_option = '<code>--enable-icmp</code>';
```

dannil replied:
That variable is for mapping the option in the rpcd output.

```
root@labmouse:~# ubus call luci.squid getOptions
{
        "options": [
                "--target=x86_64-openwrt-linux",
                "--host=x86_64-openwrt-linux",
                "--build=x86_64-pc-linux-gnu",
                "--disable-dependency-tracking",
                "--program-prefix=",
                "--program-suffix=",
                "--prefix=/usr",
                "--exec-prefix=/usr",
                "--bindir=/usr/bin",
                "--sbindir=/usr/sbin",
                "--libexecdir=/usr/lib",
                "--sysconfdir=/etc",
                "--datadir=/usr/share",
               "--enable-icmp", # Would show up here if set 
                "<many, many more options here>"
        ]
}
```

I assume you meant something like this?

```diff
diff --git a/applications/luci-app-squid/htdocs/luci-static/resources/view/squid.js b/applications/luci-app-squid/htdocs/luci-static/resources/view/squid.js
index abd2472e72..93d443d0f2 100644
--- a/applications/luci-app-squid/htdocs/luci-static/resources/view/squid.js
+++ b/applications/luci-app-squid/htdocs/luci-static/resources/view/squid.js
@@ -90,7 +90,7 @@ return view.extend({
                var enable_icmp_option = '--enable-icmp';
                var is_enable_icmp_defined = options.includes(enable_icmp_option);
                o = s.taboption('general', form.Flag, 'pinger_enable', _('Enable ICMP pinger'),
-                       !is_enable_icmp_defined ? _('Can only be set if Squid is compiled with the %s option').format(enable_icmp_option) : null);
+                       !is_enable_icmp_defined ? _('Can only be set if Squid is compiled with the <code>--%s</code> option').format(enable_icmp_option) : null);
                o.datatype = 'string';
                o.enabled = 'on';
                o.disabled = 'off';

systemcrash replied:
Ah, yes, right. But we prefer to avoid overloading i18n peeps with formatting gunk. 

Perhaps this?:
```js
			!is_enable_icmp_defined ? _('Can only be set if Squid is compiled with the %s option'.format('<code>%s</code>').format(enable_icmp_option)) : null);
```

systemcrash replied:
In any case, your way of scraping the CT options is really nice. I'm probably going to steal this for use elsewhere 😄

systemcrash commented on applications/luci-app-squid/htdocs/luci-static/resources/view/squid.js:
```diff
@@ -0,0 +1,134 @@
+
+var getOptions = rpc.declare({
+	object: 'luci.squid',
+	method: 'getOptions',
```
Can you rename this method to `getCompileTimeOptions` please, and have those changes reflect in the `luci.squid` file also?

dannil replied:
Absolutely.

systemcrash replied:
I'm really tempted to split this whole rpc method out and make it generic, so it can scrape any named executable. If you feel up to that bit, you're welcome, otherwise I can tackle it after merge.

dannil replied:
Feel free to do it after the merge, I think it's a good idea overall but a bit out-of-scope for the conversion part.

dannil replied:
Done.

systemcrash commented:
Once those are done, just squash all the commits together too, please.

systemcrash commented:
Great. Thanks @dannil
